From 1d9f8608a9ad769570b0e3671eb2953f650733f4 Mon Sep 17 00:00:00 2001
From: Robert Chiras <robert.chiras@nxp.com>
Date: Thu, 21 Dec 2017 11:06:44 +0200
Subject: [PATCH 3184/5242] MLK-17275-3: drm/bridge: nwl-dsi: Fix
 remove/detach

commit  2917040e9cdf40c5b4045c0f9932d2ac0f3311f2 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add a check in detach function, so that the mipi_dsi_host_unregister
will occur only if the host was registered.
Also, remove the unnecessary calls to host_unregister from probe and
remove functions.

Signed-off-by: Robert Chiras <robert.chiras@nxp.com>
Reviewed-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/nwl-dsi.c |   16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/bridge/nwl-dsi.c b/drivers/gpu/drm/bridge/nwl-dsi.c
index f5ab433..e56963e 100644
--- a/drivers/gpu/drm/bridge/nwl-dsi.c
+++ b/drivers/gpu/drm/bridge/nwl-dsi.c
@@ -999,10 +999,13 @@ static void nwl_dsi_bridge_detach(struct drm_bridge *bridge)
 	if (dsi->panel) {
 		drm_panel_detach(dsi->panel);
 		drm_connector_cleanup(&dsi->connector);
+		dsi->panel = NULL;
 	} else if (dsi->next_bridge) {
 		nwl_dsi_del_bridge(dsi->next_bridge->encoder, dsi->next_bridge);
+		dsi->next_bridge = NULL;
 	}
-	mipi_dsi_host_unregister(&dsi->host);
+	if (dsi->host.dev)
+		mipi_dsi_host_unregister(&dsi->host);
 }
 
 static void nwl_dsi_bridge_enable(struct drm_bridge *bridge)
@@ -1169,24 +1172,17 @@ static int nwl_dsi_probe(struct platform_device *pdev)
 	dsi->bridge.of_node = dev->of_node;
 
 	ret = drm_bridge_add(&dsi->bridge);
-	if (ret < 0) {
+	if (ret < 0)
 		dev_err(dev, "Failed to add nwl-dsi bridge (%d)\n", ret);
-		goto err_host;
-	}
 
-	return 0;
-
-err_host:
-	mipi_dsi_host_unregister(&dsi->host);
 	return ret;
-
 }
 
 static int nwl_dsi_remove(struct platform_device *pdev)
 {
 	struct nwl_mipi_dsi *dsi = platform_get_drvdata(pdev);
 
-	mipi_dsi_host_unregister(&dsi->host);
+	drm_bridge_remove(&dsi->bridge);
 
 	return 0;
 }
-- 
1.7.9.5

